src/plugin.js

```diff
@@ -9,11 +9,12 @@ import Provider from './provider';
 
 export default {
   create: (name, plugin, provider) => {
-    let originalMiddleware = plugin.middleware;
+    let originalMiddlewareGetter = plugin.middleware;
+    let originalMiddleware = null;
     const ProviderComponent = provider || Provider.create(name);
     plugin.middleware = (deps, provides) => {
-      if (originalMiddleware) {
-        originalMiddleware = originalMiddleware(deps, provides);
+      if (originalMiddlewareGetter && originalMiddleware === null) {
```

Comment:
I think this will cause other problems. For example, what happens if you create two apps and set up different dependencies? The caching here will persist across apps and you would get a plugin with middleware using the wrong set of dependencies.